### PR TITLE
[Model] Add support for Gemma 3

### LINKED ISSUE #14663: [New Model]: New models Gemma 3
### The model to consider.

Gemma 3 has a large, 128K context window, multilingual support in over 140 languages, and is available in more sizes than previous versions. Gemma 3 models are well-suited for a variety of text generation and image understanding tasks, including question answering, summarization, and reasoning.

Inputs and outputs
Input:

Text string, such as a question, a prompt, or a document to be summarized
Images, normalized to 896 x 896 resolution and encoded to 256 tokens each
Total input context of 128K tokens for the 4B, 12B, and 27B sizes, and 32K tokens for the 1B size
Output:

Generated text in response to the input, such as an answer to a question, analysis of image content, or a summary of a document
Total output context of 8192 tokens

https://huggingface.co/collections/google/gemma-3-release-67c6c6f89c4f76621268bb6d

https://storage.googleapis.com/deepmind-media/gemma/Gemma3Report.pdf

### The closest model vllm already supports.

Gemma 2

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR adds the support for [Gemma 3](https://huggingface.co/collections/google/gemma-3-release-67c6c6f89c4f76621268bb6d), an open-source vision-language model from Google.

NOTE:
* The PR doesn't implement the pan-and-scan pre-processing algorithm. It will be implemented by a followup PR cc @DarkLight1337 
* For the text-only inputs, both V0 and V1 should produce accurate outputs with good performance.
* For image inputs, only V0 implements the attention in the correct way. Gemma 3 uses bidirectional attention only for the image tokens, which is not supported efficiently by any of the current attention backends. Therefore, we temporarily use the naive PyTorch SDPA with masking tensors. This could lead to significant memory usage for long prompts (w/ images).
* For V1, we currently do not strictly follow the original attention in Gemma 3. The model still generates reasonable outputs, but this needs to be fixed to get the full accuracy.

Thanks for the help @ywang96 and @DarkLight1337!

FIX #14663

### CODE REVIEW

File: vllm/model_executor/models/gemma3_mm.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+from .interfaces import SupportsMultiModal, SupportsPP
+from .siglip import SiglipVisionModel
+from .utils import (AutoWeightsLoader, flatten_bn, init_vllm_registered_model,
+                    maybe_prefix, merge_multimodal_embeddings)
+
+logger = init_logger(__name__)
+
+# TODO(woosuk): Get these values from the model config.
+NUM_TOKENS_PER_IMAGE = 256
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/models/gemma3_mm.py:41]: Actually, other models (e.g., InternVL) also hardcode the values.
IIIUC, not all methods (e.g, `get_mm_max_tokens_per_item`) that need this value can access the config.

Maybe we can just delete the comment and hardcode the values. WDYT?

File: docs/source/models/supported_models.md [hunk 2]
```diff
@@ -937,6 +949,24 @@ For more details, please see: <gh-pr:4087#issuecomment-2250397630>
 To use Qwen2.5-VL series models, you have to install Hugging Face Transformers library from source via `pip install git+https://github.com/huggingface/transformers`.
 :::
 
+:::{note}
+Both V0 and V1 support `Gemma3ForConditionalGeneration` for text-only inputs.
```
Comment (DarkLight1337) [inline docs/source/models/supported_models.md:953]:
```suggestion
To use Gemma3 series models, you have to install Hugging Face Transformers library from source via
`pip install git+https://github.com/huggingface/transformers`. The earliest commit that supports this is [`50d3530aa04e7a7d003e6b255a98f79fd0447357`](https://github.com/huggingface/transformers/commit/50d3530aa04e7a7d003e6b255a98f79fd0447357).

Both V0 and V1 support `Gemma3ForConditionalGeneration` for text-only inputs.
```

HF side has merged the PR, so we can ask users to install it now.

File: tests/models/registry.py [hunk 3]
```diff
@@ -241,6 +241,7 @@ def check_available_online(
     "DeepseekVLV2ForCausalLM": _HfExamplesInfo("deepseek-ai/deepseek-vl2-tiny",  # noqa: E501
                                                hf_overrides={"architectures": ["DeepseekVLV2ForCausalLM"]}),  # noqa: E501
     "FuyuForCausalLM": _HfExamplesInfo("adept/fuyu-8b"),
+    "Gemma3ForConditionalGeneration": _HfExamplesInfo("google/gemma-3-4b-it"),
```
Comment (DarkLight1337) [inline tests/models/registry.py:244]:
```suggestion
    "Gemma3ForConditionalGeneration": _HfExamplesInfo("google/gemma-3-4b-it",
                                                      min_transformers_version="4.50"),
```

Comment (DarkLight1337) [reply tests/models/registry.py:244]: Avoid failing CI due to incorrect transformers version

File: vllm/model_executor/models/gemma3_mm.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+            img_pos = (input_token_ids == self.config.image_token_index)
+            img_mask[:, :, :, img_pos] += 1
+            img_mask[:, :, img_pos, :] += 1
+            global_attn_mask = torch.where(img_mask == 2, 0, global_attn_mask)
+            global_attn_masks.append(global_attn_mask)
+
+            # Create a local causal mask with sliding window (1024).
+            local_attn_mask = torch.ones_like(global_attn_mask)
+            local_attn_mask = torch.tril(local_attn_mask,
+                                         diagonal=-self.sliding_window_size)
```
Comment (ywang96) [inline vllm/model_executor/models/gemma3_mm.py:404]:
```suggestion
                                         diagonal=-self.sliding_window)
```

File: docs/source/models/supported_models.md (hunk 2 above)
Comment (WoosukKwon) [reply docs/source/models/supported_models.md:953]: I left a similar comment in `gemma3_mm.py`. Do we need it here as well?

Comment (DarkLight1337) [reply docs/source/models/supported_models.md:953]: Yeah, most users won't look at the code...

Comment (DarkLight1337) [reply docs/source/models/supported_models.md:953]: Also that branch has been merged into main already

Comment (WoosukKwon) [reply docs/source/models/supported_models.md:953]: ok. sg.

File: tests/models/registry.py (hunk 3 above)
Comment (WoosukKwon) [reply tests/models/registry.py]: Nice catch. Updated!

File: docs/source/models/supported_models.md (hunk 2 above)
Comment (WoosukKwon) [reply docs/source/models/supported_models.md:957]: Updated! I removed the error msg in `gemma3_mm.py` since users will likely get an error before loading the file (when instantiating Gemma3Config).

**Review by ywang96 — APPROVED** [review]
Fixed a few issues and I think this PR is good to be shipped. We can follow up regarding cropping in #14672 from @DarkLight1337

**Comment by devops724** [comment]
great, this mean we have Gemma 3 support in next release , thanks

**Comment by DarkLight1337** [comment]
Expediting merge as the other tests have passed in the previous build: https://github.com/vllm-project/vllm/pull/14660/commits/41110046a4a626bc492c33d6603938bd72d15fe0

**Comment by moficodes** [comment]
Do we know when the latest docker image is being published?

**Comment by WoosukKwon** [comment]
@DarkLight1337 @ywang96 Thanks for all the fixes!

**Comment by erdaltoprak** [comment]
> For V1, we currently do not strictly follow the original attention in Gemma 3. The model still generates reasonable outputs, but this needs to be fixed to get the full accuracy

Hi @WoosukKwon, can you explain the main difference between the two approaches ? Thanks :-)

**Comment by Dilesh-chouhan** [comment]
do we know that when will the docker latest image will published?

**Comment by Swipe4057** [comment]
How can I run Gemma 3 with a backend different from Xformers? It works very slowly. And it doesn't start with flashinfer.

**Comment by francis2tm** [comment]
> How can I run Gemma 3 with a backend different from Xformers? It works very slowly. And it doesn't start with flashinfer.

Having the same problem

**Comment by DarkLight1337** [comment]
Can you try setting `VLLM_USE_V1=0` to enable more backends?

**Comment by Swipe4057** [comment]
@DarkLight1337 
image
and 
image

**Comment by DarkLight1337** [comment]
The error message shows that your `--max-model-len` is too high

**Comment by Swipe4057** [comment]
@DarkLight1337 I understand that perfectly, but the Gemma3 model has a context window: Total input context of 128K tokens for the 4B, 12B, and 27B sizes, and 32K tokens for the 1B size.

**Comment by Swipe4057** [comment]
In general, try to run the model with any other backend and you'll see that it doesn't work, while Xformers is terribly slow.

**Comment by DarkLight1337** [comment]
It's supported if you install from main branch, but there might be correctness issues because their attention mask is not fully implemented in V1.

**Comment by anunknowperson** [comment]
>I think here the max model len corresponds to the sliding window length, not the total window length

max-model-len is model total size, see https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html

```
--max-model-len

    Model context length. If unspecified, will be automatically derived from the model config.

```

**Comment by anunknowperson** [comment]
> I think here the max model len corresponds to the sliding window length, not the total window length

--max-model-len is model total context size, see https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html

There are some problems with gemma 3 support currently. V0 doesn't support flashattention and context size is huge. And with V1 I wasn't able to load gptq quant.

**Comment by pietrobolcato** [comment]
> Edit: I was able to load gpt quant, but the context size is very low(<8192) in 24gb gpu
Edit: I was able to load 8k context with max-num-seq 10 and --gpu-memory-utilization 0.99, though I expected more context to fit since model is 16gb, so there should be 8 gb of free vram for context.

Could you share your code? @anunknowperson 🙏

**Comment by anunknowperson** [comment]
@pietrobolcato 

My launch string for openai server is
```bash
CUDA_VISIBLE_DEVICES=0 vllm serve ISTA-DASLab/gemma-3-27b-it-GPTQ-4b-128g --max-model-len 8192 --max-num-seqs 10 --gpu-memory-utilization=0.99
```

Probably you can get params from here for code. Engine is V1, V0 context is too big.
